### PR TITLE
EH: fix validation of delegate opcode

### PR DESCRIPTION
**Pull request by yamt**
cf. https://github.com/bytecodealliance/wasm-micro-runtime/issues/1884#issuecomment-1914000294

### CODE REVIEW

File: core/iwasm/interpreter/wasm_loader.c [hunk 1]
```diff
@@ -7128,6 +7128,30 @@ check_branch_block(WASMLoaderContext *loader_ctx, uint8 **p_buf, uint8 *buf_end,
...
+check_branch_block_for_delegate(WASMLoaderContext *loader_ctx, uint8 **p_buf,
+                                uint8 *buf_end, char *error_buf,
+                                uint32 error_buf_size)
+{
+    uint8 *p = *p_buf, *p_end = buf_end;
+    BranchBlock *frame_csp_tmp;
+    uint32 depth;
+
+    read_leb_uint32(p, p_end, depth);
+    CHECK_BR(depth);
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_loader.c:7142]: Not sure whether it should be `CHECK_BR(depth + 1)`? Seems the target block to check is `loader_ctx->frame_csp - depth - 2`.
In check_branch_block, the code is:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/3111a86cb341cee9eb0b3e7d8d1d43327846a627/core/iwasm/interpreter/wasm_loader.c#L7120-L7121

Comment (yamt) [reply core/iwasm/interpreter/wasm_loader.c:7142]: good point. i will check.

Comment (yamt) [reply core/iwasm/interpreter/wasm_loader.c]: fixed.

File: core/iwasm/interpreter/wasm_loader.c [hunk 2]
```diff
@@ -7129,6 +7129,40 @@ check_branch_block(WASMLoaderContext *loader_ctx, uint8 **p_buf, uint8 *buf_end,
...
+
+    read_leb_uint32(p, p_end, depth);
+    /*
+     * Note: "delegate 0" means the surrounding block, not the
+     * try-delegate block itself.
+     *
+     * Note: the caller hasn't popped the try-delegate frame yet.
+     */
+    bh_assert(loader_ctx->csp_num > 0);
+    if (loader_ctx->csp_num - 1 <= depth) {
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_loader.c:7150]: Not sure whether the check is enough? The original code calls check_branch_block:
check_branch_block -> CHECK_BR -> wasm_loader_check_br
And in wasm_loader_check_br, there are more checks:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/3111a86cb341cee9eb0b3e7d8d1d43327846a627/core/iwasm/interpreter/wasm_loader.c#L7058-L7088

Is it better to use `CHECK_BR(depth + 1)`?

Comment (yamt) [reply core/iwasm/interpreter/wasm_loader.c:7150]: the most of checks in wasm_loader_check_br are not appropriate for delegate, where parameters are given via the exception, not stack.

Comment (wenyongh) [reply core/iwasm/interpreter/wasm_loader.c:7150]: OK, got it, thanks!

**Review by wenyongh — APPROVED** [review]

**Pull request merged**